paas_app_charmer/utils.py

```diff
@@ -51,3 +55,52 @@ def enable_pebble_log_forwarding() -> bool:
         return True
     except ImportError:
         return False
+
+
+@functools.lru_cache
+def _config_metadata(charm_dir: pathlib.Path) -> dict:
+    """Get charm configuration metadata for the given charm directory.
+
+    Args:
+        charm_dir: Path to the charm directory.
+
+    Returns:
+        The charm configuration metadata.
+
+    Raises:
+            ValueError: if the charm_dir input is invalid.
+    """
+    config_file = charm_dir / "config.yaml"
+    if config_file.exists():
+        return yaml.safe_load(config_file.read_text())
+    config_file = charm_dir / "charmcraft.yaml"
+    if config_file.exists():
+        return yaml.safe_load(config_file.read_text())["config"]
+    raise ValueError("charm configuration metadata doesn't exist")
+
+
+def config_get_3(
```

Comment:
Yes, this is intended to be a placeholder for the functionality to be merged into the `ops` framework.

Reply:
Yes, I felt that `Model.config` should give you a `Secret` not a `str` of a secret ID, because it's pretty worthless without always following it up with `get_secret`). There were arguments that it's too much of a divergence from the 'simple' types that `config` currently has, and I was also slightly late in my realisation/argument, so that didn't happen.

I still think ops should indeed do something with this. It's mentioned in https://github.com/canonical/operator/issues/1223 - not the primary focus of the ticket, but as something that seems likely to happen in it. Charm-Tech has work in this area tentatively scheduled for 25.04.

It is likely that an ops equivalent of this would return the secret object, not also get the content and return that, so that we aren't pulling in sensitive content when it might not get used (and also avoiding having to decide whether to refresh). But it would still make this much simpler.

Thanks for the ping! References to real-world use-cases are always very useful both when considering whether to implement a feature, and then in the actual design & implementation.

As an aside: +1 on renaming this :)

Reply:
Great thanks, so what do you think about renaming the function please Weii?

Reply:
The `config_get_3` means config get (juju)3. Perhaps we could change it to `config_get_with_secret`?

Reply:
yes that name sounds good!

Reply:
Updated, thanks!